### PR TITLE
fix:[DPMMA-2636] fix reports page hits - issue 13485

### PR DESCRIPTION
**Pull request by AlanWierzchonCA**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [yes]
| New feature/enhancement? (use the a.x branch)      | [No]
| Deprecations?                          | [No]
| BC breaks? (use the c.x branch)        | [No]
| Automated tests included?              | [No] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

Fix Issue: https://github.com/mautic/mautic/issues/13485

### CODE REVIEW

**Review by patrykgruszka — COMMENTED** [review]
I'm wondering why there are records with empty values. Can you please check the file `app/bundles/PageBundle/EventListener/ReportSubscriber.php` at line 539 and the `HitRepository::getMostVisited` method? We might need to fix the query if there's a problem.

The indentation in the twig file needs to be adjusted. The `<tr>` element seem to be out of position.

**Comment by AlanWierzchonCA** [comment]
I added a Where clause to the query in HitRepository::getMostVisited.

**Review by patrykgruszka — APPROVED** [review]
Looks good now, thanks!

File: app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig [hunk 1]
```diff
@@ -21,25 +21,27 @@
...
+                                            </td>
+                                        {% elseif 'id' != cellName %}
+                                            <td>
+                                                {% if cell is numeric %}
+                                                    {{ cell }}
+                                                {% else %}
+                                                    {{ assetMakeLinks(cell) }}
+                                                {% endif %}
+                                            </td>
                                         {% endif %}
```
Comment (escopecz) [inline app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig:41]: I don't quite understand this change. The `assetMakeLinks` function is called only if the `cellName` is NOT equal to `id`. So it cannot be the `id` part that is `null`, right? Can you provide a functional test that would show the data that are causing the error in the linked issue?

Comment (AlanWierzchonCA) [reply app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig:41]: Sometimes there is a pagehits record with pageId equal to null. This produces null results and a 500 error.

Comment (AlanWierzchonCA) [reply app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig:41]: @escopecz, I added functional tests.

File: app/bundles/PageBundle/Tests/Model/PageModelFunctionalTest.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        $hit->setIpAddress(new IpAddress('127.0.0.1'));
+
+        $this->em->persist($hit);
+        $this->em->flush();
+
+        $query = $this->em->getConnection()->createQueryBuilder();
+        $query->from(MAUTIC_TABLE_PREFIX.'page_hits', 'ph');
+        $query->leftJoin('ph', MAUTIC_TABLE_PREFIX.'pages', 'p', 'ph.page_id = p.id');
+
+        $res = $this->pageModel->getHitRepository()->getMostVisited($query);   // $this->em->getRepository(Hit::class);
```
Comment (escopecz) [inline app/bundles/PageBundle/Tests/Model/PageModelFunctionalTest.php]:
```suggestion
        $res = $this->pageModel->getHitRepository()->getMostVisited($query);
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Looks good! The first 25 lines of both tests are the same. Do you think this code could be in 1 place and reused twice?

**Comment by AlanWierzchonCA** [comment]
> Looks good! The first 25 lines of both tests are the same. Do you think this code could be in 1 place and reused twice?

@escopecz I changed it, as you suggested.

**Review by escopecz — APPROVED** [review]
Looks good. Thank you! I was able to reproduce the issue. This is fixing it and covering with tests 👍

File: app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig [hunk 3]
```diff
@@ -21,25 +21,27 @@
...
-                                            {{ cell }}
-                                        </a>
-                                    </td>
-                                {% elseif 'id' != cellName %}
-                                    <td>
-                                        {% if cell is numeric %}
-                                            {{ cell }}
-                                        {% else %}
-                                            {{ assetMakeLinks(cell) }}
+                            {% if row['id'] is not null %}
```
Comment (escopecz) [inline app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig:24]: This caused a bug: https://github.com/mautic/mautic/issues/13695

Can you please make a fix? I think you have to check if the ID key defined first like on line 27

Comment (AlanWierzchonCA) [reply app/bundles/ReportBundle/Resources/views/Graph/Table.html.twig:24]: ok, I will.

**Review by obamwonyi — COMMENTED** [review]
I think line 24 is failing because the key 'id' in the variable row['id'] is missing. Therefore, if the key 'id' is not present, the interpreter will still encounter an error. In simpler terms, we need to find a way for the interpreter not to fail even if it can't locate the 'id' key.
